rust-runtime/smithy-json/src/serialize.rs

```diff
@@ -110,49 +138,41 @@ impl<'a> JsonArrayWriter<'a> {
         }
     }
 
-    /// Writes a null value to the array.
-    pub fn null(&mut self) -> &mut Self {
+    #[inline]
+    fn write<F: Fn(JsonValueWriter) -> ()>(&mut self, f: F) -> &mut Self {
         self.comma_delimit();
-        self.json.push_str("null");
+        f(JsonValueWriter::new(&mut self.json));
         self
     }
 
+    /// Writes a null value to the array.
+    pub fn null(&mut self) -> &mut Self {
+        self.write(|w| w.null())
+    }
+
     /// Writes the boolean `value` to the array.
     pub fn boolean(&mut self, value: bool) -> &mut Self {
```

Comment:
Since the JsonValueWriter consumes itself every time it writes a value, I was trying to avoid this awkward interface:
```rust
array.value().string("1").value().string("2")
```
I suppose this doesn't really matter for a codegen though, since it would only ever generate something that looked like:
```
for item in &some_struct.some_list {
   array.value().string(item);
}
```
So, given that, I think I'm onboard with cutting this down.